libs/features/japanese-word-wrap.js

```diff
@@ -1,27 +1,214 @@
+ * @returns true if the element contains a non-empty text node, false otherwise.
  */
-export async function controlLineBreaksJapanese(config, area = document, selector = 'h1, h2, h3, h4, h5, h6') {
+function hasTextNode(element) {
+  for (let i = 0; i < element.childNodes.length; i += 1) {
+    const child = element.childNodes[i];
+    if (child.nodeType === Node.TEXT_NODE && child.textContent.trim() !== '') {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+/**
+ * Recursively finds all elements that contain non-empty text nodes.
+ * The search stops when an element with the 'jpwordwrap-disable' class is encountered.
+ *
+ * @param {HTMLElement} [element=document.body] - The root HTMLElement to start the search from.
+ * @returns An array of DHTMLElements that contain non-empty text nodes.
+ */
+function findTextElements(element = document.body) {
+  let result = [];
+
+  if (
+    element.classList.contains('jpwordwrap-disabled')
+    || element.tagName.toLowerCase() === 'header'
+    || element.tagName.toLowerCase() === 'footer'
+  ) {
+    return [];
+  }
+
+  for (let i = 0; i < element.children.length; i += 1) {
+    const child = element.children[i];
+    if (hasTextNode(child)) {
+      result.push(child);
+    } else {
+      result = result.concat(findTextElements(child));
+    }
+  }
+
+  return result;
+}
+
+/**
+ * Update the model to control line breaks occurring for the specified word.
+ * @param {*} parser The BudouX parser to update.
+ * @param {string} pattern The pattern that should be updated in the model.
+ * @param {number} score The score that should be added from the pattern in the model.
+ * @param {string} [markerSymbol='#'] The marker symbol to look for in the pattern string.
+ * Defaults to '#'.
+ */
+function updateParserModel(parser, pattern, score, markerSymbol = '#') {
+  const markerPos = pattern.indexOf(markerSymbol);
+  if (markerPos === -1) {
+    console.warn('No marker symbol found in the line break pattern string');
+    return;
+  }
+
+  if (markerPos !== pattern.lastIndexOf('#')) {
+    console.warn('Two or more marker symbols cannot be specified. Only the first marker is applied');
+  }
+
+  const former = pattern.slice(Math.max(markerPos - 3, 0), markerPos);
+  const latter = pattern.slice(markerPos + 1, Math.min(markerPos + 4, pattern.length));
+
+  if (former.length < 2 || latter.length < 2) {
+    console.warn('At least two characters must be specified before and after the marker symbol');
+    return;
+  }
+
+  if (former.length === 3) {
+    parser.model.set(`TW1:${former}`, score);
+  } else if (former.length === 2) {
+    parser.model.set(`BW1:${former}`, score);
+  }
+
+  if (latter.length === 3) {
+    parser.model.set(`TW4:${latter}`, score);
+  } else if (latter.length === 2) {
+    parser.model.set(`BW3:${latter}`, score);
+  }
+}
+
+/**
+ * Get metadata
+ * @param {*} name The name of metadata.
+ * @returns metadata content
+ */
+function getMetadata(name) {
```

Comment:
you can import this from utils

Reply:
@chrischrischris I wanted to, but it would be a circular reference.

Reply:
https://github.com/adobecom/milo/blob/main/libs/utils/utils.js#L748-L750

Reply:
We've disabled the circular-ref error/warning since in the majority of cases importing from utils is not an issue, so go ahead and import it.